### ISSUE TITLE
[Proposal]: blind accessibility

### ISSUE THREAD

**Issue by bobbens** [naev/naev#2635]
### Suggestion

Adding blind accessibility (not colourblind, full blindness) could be a good goal at not only improving the accessibility for blind users, but for all users in general. I've opened an issue to try to centralize information regarding the topic and possible ways of handling it. However, this is more of a mid/long-term goal than something we can realistically do in the short term.

So while for pure C, there seems to be no really good options, there seems to be a lot of libraries we can use written in rust:
* [AccessKit](https://github.com/AccessKit/accesskit): can be used to wrap around the GUI and make it more accessible. I don't think it's specific for blind users, so many more could benefit. Uses cbindgen to generate c bindings too. Does seem still in development, so I'm not sure how usable it is.
* [tts-rs](https://github.com/ndarilek/tts-rs): text to speech support that is cross-platform. Seems quite solid.

What is probably easiest to get working is the Visual Novel framework, which would just need some tts support. Toolkit shouldn't be too hard either with AccessKit, assuming it works well enough. That leaves in-game play, which is probably going to be a much harder issue to tackle. Not entirely sure the best way to handle that right now.

Alternatively, it may be a better long-term project to switch to some GUI toolkit that does this automatically. However, that may be a bit out of the scope for now. [egui](https://github.com/emilk/egui?tab=readme-ov-file) (also rust) seems like a good contender for now.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Zireael07** [comment]
Godot has an open PR that adds AccessKit support. No clue why it hasn't been merged yet.

Considering Naev is a real time space game, I have ZERO clue where you would even start with making the actual gameplay accessible. I know people made Minecraft accessible but that is voxel- and grid-based

**Comment by patricus3** [comment]
you also should add some kind of nearest objects and possibility to navigate to it, and a map with other places that are further away

**Comment by zersiax** [comment]
For the gameplay sections a number of things come to mind:

- Provide persistent audio cues for currently visible objects (e.g. planets) that are directional based on current perspective. That way, sound can be used to steer towards these objects.
- For thingsthat are further away, I'm assuming these canbe visually "tracked" somehow. We'd need to come up with an audio/TTS equivalent for that.
- I don't necessarily think you need a UI toolkit, that is actually pretty uncommon in the videogame sphere. Long as you can send the currently highlighted menu item, as well as the currently scrolled/popped up/etc. text to some kind of TTS renderer, you have working menus, dialogues, lore readouts and alerts. For menu items it could be as simple as firing that tts event on focusing the rquisite menu item; sometimes this is an onHover event but generally the blind don't use a mouse so that would only work if controller/keyboard focus triggers that event.

Some food for thought perhaps.

**Comment by Zireael07** [comment]
For things further away, a map/menu thingy would work. Cataclysm:DDA has such a menu that lists all the objects on screen with directions and distances

**Comment by bobbens** [comment]
> Honestly what I personally like is as few differences as possible for accessibility. Instead of trying to add in all of these custom things maybe finding a way to make current systems that are in place work better in terms of accessibility. This will reduce dev work and make things easier to maintain. Starting with the menus is definitely the best place to start and at least then a blind person like me could at least try the game and give feedback on how things could be improved. 

I totally agree. For mos of the GUI and Visual Novel (VN) interfaces, I don't think it'll be an issue at all. All we have to do is use AccessKit + have the VN just connect to TTS stuff and have the characters talk. Honestly, if it were possible, I would love to have fully voiced characters, but there's no way that is feasible.

> Maybe having a map of sorts that has all of the possible highlightable things in an area would also be good and not just for blind players. It could contain filters or an easy way to look at particular things like inhabited planets, trading ships, etc. I'm not sure if this is helpful but if I could get a better understanding of the game then I could give better advice.

The map has a few modes, like showing fully explored systems, commodity prices, but things like inhabited planets and trade routes are shown using symbols. Also your reputation in the system, and available services, but you have to select a system to see that (although you can see if they're hostile / neutral / friendly on the map without clicking).

There is also support for finding / searching for space objects (planets / stations), and systems, so with accesskit hooked up to the GUI, that "should" work out of the box. In the most part, I think hooking up the custom widget that is the map to accesskit should probably work to some extent.

The main issue is going to be the gameplay itself. as it is real-time and there is a lot of information to convey, so there will have to be some sort of simplification or prioritization. In the game, you basically go from star system to star system, which act as discrete areas. You can only see the inside of the star system you are in. In a single star system, you usually have something like:

>     1. A couple of inhabited worlds. They are fixed location and can be interesting.
>     2. A couple of uninhabited words. Mainly for background, probably can be ignored.
>     3. A couple of jump points. They are necessary for traveling. They can probably be ignored most of the time except when opening the map or autonaving through them.
>     4. 10-50 ships. Most will not be interesting, and just be traders or independents flying around. However, some may be more interesting such as pirates or enemies.
>     5. Neutral or enemy fire such as torpedoes, missiles, long-range artillery and such. This can be very noisy, especially for fast kinetic weapons.

You can select independently ships, space objects (planets / stations),  jumps, and asteroids. The game also depends on an "automatic navigation" or autonav system, that can automatically let you go to locations, follow ships, land on space objects, or go to some remote system by going to the jumps and jumping. When autonav is on, the game time accelerates, so it's not that important to convey information.

I think something that could make sense would be to have like audio location indicators for your targets. Probably different by type. So if you target a ship and a space object, you would get the relative location to your ship and be able to use that to steer or whatever. You can change targets to switch the audible signal and use the cycling keys to see what's in the system. It could also send the ship name and more info to the TTS as you change targets. This might be enough for the bare minimum to be able to play the game.

Currently the game shows 2D sound using fixed coordinates. As in sounds on the left are on the left of the display, regardless to how the ship is facing. However, back in the day, it was actually relative to the ship. As in stuff on the left of the direction the ship is facing sounded to be on the left. It should be trivial enough to add an accessibility mode to reenable that old behaviour, and that could also enable the TTS stuff and other accessibility things.

Implementation-wise, it shouldn't be "too hard" once the GUI issue is solved, and AccessKit + TTS is interfaced with Naev. For gameplay, pretty much all we would have to do would be to add some Lua API for AccessKit + TTS, and then have some event that starts when the player loads a pilot (or starts a new game) that would basically sit in the background and detect when the player changes targets, and play sounds and the likes. All the pilot / target / spob information is completely accessible via Lua, so this would be a good way to make it flexible and easy to use.

Other things that I can think of is we'll need a lot more audible indicators for different things. For example, there are no audible indicators when outfits finish cooling down or weapons are overheating. However, it would make sense to try to get the minimum stuff needed to play without relying on visuals first, and then we can probably iron out the small issues as they come by.

Probably the first things that need to decided before doing other things are:
1. What to do with the GUI?  I'm partial to seeing if the [iced](https://iced.rs/) fits the purpose, and they should get AccessKit before we can implement it, so that should solve things.
2. What library to use for TTS? Again a [tts-rs](https://docs.rs/tts/latest/tts/) library in rust seems the best option, however, I would like an opinion of an actual user of it if possible.
3. Do we have to do anything other than AccessKit + TTS and the in-game sound indicators? Do they interplay well with each other?

Naev 0.12.0 is going to go into feature freeze in about a month or less, and rust is not going to be part of 0.12.0, so all of this is basically going to be part of next year's development. Furthermore, the Rust stuff, although working, is still experimental, and we have to see if it compiles / links under different operating system combinations and the likes. However, that should be less of an issue.

Sorry for the long rambling.

**Comment by paulGeoghegan** [comment]
> Probably the first things that need to decided before doing other things are:
> 
>     1. What to do with the GUI?  I'm partial to seeing if the [iced](https://iced.rs/) fits the purpose, and they should get AccessKit before we can implement it, so that should solve things.

I definitely think this is probably the best idea and will probably work quite well.

>     2. What library to use for TTS? Again a [tts-rs](https://docs.rs/tts/latest/tts/) library in rust seems the best option, however, I would like an opinion of an actual user of it if possible.

It seems like this supports TTS directly to a screen reader but only on windows. It is cross platform and seems pretty good. I couldn't find any examples of how it actually sounds but you would probably need to add controls for changing the TTS options in the game.

>     3. Do we have to do anything other than AccessKit + TTS and the in-game sound indicators? Do they interplay well with each other?

I think that is probably all you will need. I could be wrong but at least when these things are done I could play test it and give better feedback.
Honestly some accessibility is better than none and this is the best place to start.

Definitely don't be sorry and again thank you for your interest.

**Comment by DataTriny** [comment]
Hello, AccessKit developer here.

While the library is still in active development, we already support all the widget types you should ever need in a game UI, on all desktop platforms.

In fact we already demonstrated AccessKit's capabilities in a game environment with [this adaptation of a game made with Unity](https://github.com/accesskit/the-intercept). We didn't use our C bindings as they didn't exist back then, instead going through a simpler C API and JSON serialization of the tree updates.

Here is [a video showing the end result](https://youtu.be/m9UkOuXu6Z4?t=444) recorded during NarraScope 2022.

The story is delivered to the screen reader user by putting the dialogues in something known as a "live region". Depending on your use case, this might be enough. Libraries such as tts-rs will only give you slightly more flexibility.

As a visually impaired myself, I might give this game a try. I'm available if you have more questions.

**Comment by bobbens** [comment]
> While the library is still in active development, we already support all the widget types you should ever need in a game UI, on all desktop platforms.

Just so I'm clear on this. Does AccessKIt cover all the things a TTS library would do? Or would we need to use both?

Some other questions:
1. You do not have a meson script (or meson wrap) for AccessKit, do you? On our [rust](https://github.com/naev/naev/tree/rust) we have a fairly complex build system where we use meson to build a C library for Naev, cargo to build a rust library, then meson to compile the C library and rust library using meson for dependencies. We can figure out something if there isn't, however, it will take a bit more time.
2. If we use something like [iced](https://iced.rs/), we would still be able to use AccessKit separately I guess?
3. I'm trying to find some documentation on where to get started. Is there some sort of tutorial or simple guide to get started for either rust or C? I can base it off the SDL / C example, but it's likely we'll be using rust to initialize, and probably also use the C interface on the C side, at least until most stuff gets replaced with rust.
4. Is there any API for giving 2D/3D spatial audio hints? or would we have to do that ourselves with OpenAL?
5. I assume it would work well with translations? We can just send the Japanese UTF-8 text or whatever and it'll handle it?
6. It seems to focus on winit with Rust, would it be possible to use it with SDL as with the C example?

> As a visually impaired myself, I might give this game a try. I'm available if you have more questions.

That would be great! Probably would be worth getting some sort of proof of concept first.

**Comment by DataTriny** [comment]
> Just so I'm clear on this. Does AccessKIt cover all the things a TTS library would do? Or would we need to use both?

AccessKit allows you to describe your UI to the operating system in a standardized way, TTS libraries are just a way to output voice. You could try to simulate an accessible UI using a TTS library but that would feel hacky to users (although this is what most audio games do today). With a powerful TTS API you can do stuff like pausing or cancelling speech and be notified when the speech synthesizer is done speaking text so you can play sounds or trigger other game actions. I'd recommend going the more generic way with AccessKit and see if you need a TTS at some point.

> 1. You do not have a meson script (or meson wrap) for AccessKit, do you?

There is [an open PR](https://github.com/AccessKit/accesskit-c/pull/2) for that.

> 2. If we use something like iced, we would still be able to use AccessKit separately I guess?

Iced would have to give you a way to create your own widgets and provide accessibility-related attributes for them, or letting you create your own AccessKit nodes and incorporate them in the whole accessibility tree. This will have to be figured out once they make progress with their AccessKit integration.

> 3. I'm trying to find some documentation on where to get started. Is there some sort of tutorial or simple guide to get started for either rust or C?

> 4. Is there any API for giving 2D/3D spatial audio hints? or would we have to do that ourselves with OpenAL?

AccessKit only concerns itself with representing your UI in the accessibility tree, so you will have to do this yourself.

> 5. I assume it would work well with translations?

Yes, we use UTF-8 strings everywhere. We might still miss stuff regarding RTL in some places though.

> 6. It seems to focus on winit with Rust, would it be possible to use it with SDL as with the C example?

It seems totally doable to me using [these SDL bindings](https://github.com/Rust-SDL2/rust-sdl2).

**Comment by bobbens** [comment]
> > Just so I'm clear on this. Does AccessKIt cover all the things a TTS library would do? Or would we need to use both?
> 
> AccessKit allows you to describe your UI to the operating system in a standardized way, TTS libraries are just a way to output voice. You could try to simulate an accessible UI using a TTS library but that would feel hacky to users (although this is what most audio games do today). With a powerful TTS API you can do stuff like pausing or cancelling speech and be notified when the speech synthesizer is done speaking text so you can play sounds or trigger other game actions. I'd recommend going the more generic way with AccessKit and see if you need a TTS at some point.

Maybe it's me not understanding this. From the demo video you linked, it seems like it is outputting voice and the likes with AccessKit, so I would assume that AccessKit is describing the UI to the operating system, which then resorts to TTS or the likes? Would it be possible to do this with game elements or would it be better to use the TTS API directly. Furthermore, do they conflict? As in, if AccessKit is describing something and outputting sound, and you tell the TTS to do the same, would they overlap?

I'm thinking about a potential case in-game. There will be an in-game GUI that will likely be using AccessKit (this is independent of iced and is written in Lua). Furthermore, as discussed a bit before, things like changing targets by hotkey would have to also output short audio descriptions. In this case, would that be possible to do with AccessKit or would that need the TTS?

Similarly, for the visual novel, which is just some simple interface that constantly streams text and the player hits space to continue the story. Would this be possible to do with AccessKit or would we need the TTS support? I guess in this case, if the TTS library supported different characters, we could use different voices to try to convey more information easily.

> > 1. You do not have a meson script (or meson wrap) for AccessKit, do you?
> 
> There is [an open PR](https://github.com/AccessKit/accesskit-c/pull/2) for that.

Cool!

> > 3. Is there some sort of tutorial or simple guide to get started for either rust or C?
> 
> Not at this point unfortunately. We have the examples programs you probably already found, and a lot of the node's properties follow the ARIA specification, so you can use this material to learn how to represent your UI.

OK. Yeah, it's a bit much to wrap my head around atm. I'll have to give it a more serious look once we finish the current Naev release next year.

> > 6. It seems to focus on winit with Rust, would it be possible to use it with SDL as with the C example?
> 
> It seems totally doable to me using [these SDL bindings](https://github.com/Rust-SDL2/rust-sdl2).

I'm a bit confused because looking at the [crates.io accesskit package](https://crates.io/crates/accesskit), it mentions you need a platform adapter and only lists winit and windows. I would assume this would be something we would have to write ourselves?

Thanks for the prompt replies.

**Comment by DataTriny** [comment]
> From the demo video you linked, it seems like it is outputting voice and the likes with AccessKit, so I would assume that AccessKit is describing the UI to the operating system, which then resorts to TTS or the likes?

Exactly. The voice you hear in the video is produced by a screen reader which uses its own TTS library to manipulate a speech synthesizer. AccessKit exposes the game content to the operating system's accessibility stack and the screen reader listens for changes in the accessibility tree to determine when things worthy of interest happen on the desktop.

> Furthermore, do they conflict?

This will depend on which components are involved, which will vary depending on the platform. But probably yes and you won't be able to detect it from the game code.

Apart from that, all the use cases you describe can be handled by AccessKit. You just have to be aware that there is one accessibility tree per window. So it doesn't matter who is producing AccessKit nodes, you will have to merge everything in the end. So if you end up using Iced, you will have to find a way for your Lua code to push its AccessKit nodes to the Iced tree, because they are inside the same window.

> I guess in this case, if the TTS library supported different characters, we could use different voices to try to convey more information easily.

Except TTS libraries won't usually offer this level of control. If they do, it is only available for a limited number of speech synthesis backends. If this is what you need, I would suggest talking to a speech synthesizer library directly.

> I'm a bit confused because looking at the crates.io accesskit package, it mentions you need a platform adapter and only lists winit and windows. I would assume this would be something we would have to write ourselves?

Yes. If you use SDL then you will have to create your own adapter, taking inspiration from both our SDL example and our winit adapter. If you want to do it in C today, the SDL example is fully functional so you would just have to copy it in your codebase. Would it help you if I write an SDL adapter in Rust? Do you plan to replace SDL by winit during the Rust transition?

**Comment by bobbens** [comment]
Thanks for the replies.

> > Furthermore, do they conflict?
> 
> This will depend on which components are involved, which will vary depending on the platform. But probably yes and you won't be able to detect it from the game code.

OK. Regarding this, I'm not sure how to approach it. If AccessKit can cover all our needs, that might be easiest and lead to least amount of issues. The only downside is not being able to use different voices for different characters which a TTS library may offer.

> Apart from that, all the use cases you describe can be handled by AccessKit. You just have to be aware that there is one accessibility tree per window. So it doesn't matter who is producing AccessKit nodes, you will have to merge everything in the end. So if you end up using Iced, you will have to find a way for your Lua code to push its AccessKit nodes to the Iced tree, because they are inside the same window.

OK. Is this something that can be realistically be done? I'm not sure what the iced API will look like, but assuming it's similar to other GUIs, is it feasible to use it and merge it with other custom things and Lua and whatnot? I guess the most problematic will be Lua since it'll have to use some sort of interface to obtain whatever AccessKit is doing, likely leading to some sort of singleton paradigm.

> > I guess in this case, if the TTS library supported different characters, we could use different voices to try to convey more information easily.
> 
> Except TTS libraries won't usually offer this level of control. If they do, it is only available for a limited number of speech synthesis backends. If this is what you need, I would suggest talking to a speech synthesizer library directly.

OK. I'm right to assume that AccessKit will not include this potentially in a future version? As in, if we want such a feature we have no option but to use a TTS backend directly?

> > I'm a bit confused because looking at the crates.io accesskit package, it mentions you need a platform adapter and only lists winit and windows. I would assume this would be something we would have to write ourselves?
> 
> Yes. If you use SDL then you will have to create your own adapter, taking inspiration from both our SDL example and our winit adapter. If you want to do it in C today, the SDL example is fully functional so you would just have to copy it in your codebase. Would it help you if I write an SDL adapter in Rust? Do you plan to replace SDL by winit during the Rust transition?

I don't think I will bother with doing it in C now. Regarding moving to SDL to winit. I'm not sure actually. I was pretty dead set on just sticking to SDL, which I'm quite fond of, instead of having to resort to a lot of different crates to cover same functionality. Even if we make a switch, it'll probably be quite down the line, so having an SDL adapter in the meantime would likely be necessary.

**Comment by Zireael07** [comment]
Seconding the other comment that I'm not aware of a TTS library that offers different character voices.

**Comment by bobbens** [comment]
> Seconding the other comment that I'm not aware of a TTS library that offers different character voices.

tts-rs seems to support it.
https://docs.rs/tts/latest/tts/struct.Voice.html

**Comment by DataTriny** [comment]
> I'm not sure what the iced API will look like, but assuming it's similar to other GUIs, is it feasible to use it and merge it with other custom things and Lua and whatnot?

The simplest option will probably be to have some Rust code responsible for building your regular and custom widgets, then exposing this functionality to your Lua sandbox. But I know that game developers are reluctant to use general purpose UI toolkits as they aren't usually flexible enough for stuff you want in a game. Still if you want to go with Iced, we'll have to make sure it is possible to create AccessKit nodes for custom widgets.

> OK. I'm right to assume that AccessKit will not include this potentially in a future version? As in, if we want such a feature we have no option but to use a TTS backend directly?

Such feature is out of scope for AccessKit.

> tts-rs seems to support it.

tts-rs is a generic interface to various speech synthesizers and screen reader APIs. It exposes a `Voice` struct because some backends expose such details themselves, but probably half of them don't.

> so having an SDL adapter in the meantime would likely be necessary

I have already started working on it anyway. I am not sure however if Iced is capable of running inside SDL, and if so we'll have to see how Iced and this SDL AccessKit adapter can work together.

**Comment by Zireael07** [comment]
What the other comment said. Most TTS libraries will allow you to pick gender and language, but usually work poorly for anything non-English. (Usually just adding some sort of accent that ends up being hilarious for people who actually speak the language). (I spent a lot of time twiddling with Oolite's TTS settings - it connects to the system's TTS IIRC - before realizing that no matter what I did I wouldn't get actual Polish output)

Which means they'll totally fail if you have two females talking in the story. They'll both need to use the same setting(s)

**Comment by bobbens** [comment]
> > I'm not sure what the iced API will look like, but assuming it's similar to other GUIs, is it feasible to use it and merge it with other custom things and Lua and whatnot?
> 
> The simplest option will probably be to have some Rust code responsible for building your regular and custom widgets, then exposing this functionality to your Lua sandbox. But I know that game developers are reluctant to use general purpose UI toolkits as they aren't usually flexible enough for stuff you want in a game. Still if you want to go with Iced, we'll have to make sure it is possible to create AccessKit nodes for custom widgets.

Yes, before we commit I'll try to do a proof of concept and evaluate how good it may work. Worst case, we'll just rewrite ours in Rust and add AccessKit, but it would be nice if we could outsource.

> > tts-rs seems to support it.
> 
> tts-rs is a generic interface to various speech synthesizers and screen reader APIs. It exposes a `Voice` struct because some backends expose such details themselves, but probably half of them don't.

> hat the other comment said. Most TTS libraries will allow you to pick gender and language, but usually work poorly for anything non-English. (Usually just adding some sort of accent that ends up being hilarious for people who actually speak the language). (I spent a lot of time twiddling with Oolite's TTS settings - it connects to the system's TTS IIRC - before realizing that no matter what I did I wouldn't get actual Polish output)
> 
> Which means they'll totally fail if you have two females talking in the story. They'll both need to use the same setting(s)

I see. So it may be best to just give up on that idea in the first place and rely on AccessKit instead.

> > so having an SDL adapter in the meantime would likely be necessary
> 
> I have already started working on it anyway. I am not sure however if Iced is capable of running inside SDL, and if so we'll have to see how Iced and this SDL AccessKit adapter can work together.

Cool! Looking forward to it.

Regarding Iced, it technically seems possible according to https://github.com/iced-rs/iced/tree/master/examples/integration . However, I'll probably write a toy case first to explore the possibilities and see if something like sdl + wgpu + accesskit + iced can work.

**Comment by azurejoga** [comment]
Accessibility can be done with compatibility for a screen reader instead of a sapi voice or a tts voice, since many are not English speakers and with the screen reader we can translate the game with translation add-ons from the screen readers themselves.

To track objects and NPCS among other things, the name of the item must be announced to the screen reader, followed by the beep that shows how far away it is from the player, that would be incredible.
The name of the object should be announced when you are close to it, this means a 3 to 5 steps close.
Also, I discourage adoption through multiple speakers and through tts, as this would discourage a number of people from playing the game, and yes. one voice is enough.
Furthermore, the navigation system could be made where the player marks where he wants to go/track and a continuous beep indicates the direction in which he should go! 
Thank you for listening to my suggestions, and below I leave some libraries for implementation for the screen reader
[tolk](https://github.com/dkager/tolk)
[Universal Speech](https://github.com/qtnc/UniversalSpeech)
[SRAL](https://github.com/m1maker/SRAL)

**Comment by DataTriny** [comment]
@azurejoga The libraries you've listed are basically accomplishing the same thing as tts-rs discussed above, in fact tolk is the first backend listed on tts-rs's README.

Calling screen reader APIs is basically text-to-speech, with primitive Braille support, maybe the ability to cancel speech and no risk of overlap between speech sequences as the screen reader is in charge of everything. You can achieve exactly the same functionality (except for the speech cancellation) by using an AccessKit live region. But properly implementing the UI with AccessKit will provide a far superior experience everywhere else. And AccessKit accommodates more than screen reader users. It will not magically make the game accessible on its own, but it can allow other types of interactions such as voice control, screen magnification and whatever the user's system is configured to do.

**Comment by ndarilek** [comment]
tts-rs author here.

Seems like an interesting game. I too am blind so can't offer a lot of specific advice, but I'll give one caveat.

In general I do prefer UI accessibility through something like accesskit, but there's one situation where you'll likely want self-voicing depending on your gameplay. If any of your ship/gameplay controls are real-time--using arrows to steer a ship and other keys to engage controls for instance--those keypresses will usually interrupt the screen reader. So if someone is holding left-arrow to bank left, and an alert pops up, that alert might get skipped by the screen reader. Likewise if one appears at or near the moment when someone presses a key to shoot or perform some other twitch-based fast action. I'm making a game myself and this has been a fairly huge issue. I have to put NVDA in sleep mode to play, which suspends its handling of accessibility events and relies on the application's own self-voicing, which mine does. I don't know of a good hybrid approach so it'll be interesting if you manage to figure one out here, and I'm happy to help test/develop/brainstorm.

It's *possible* that an assertive live region might work around some or all of this, but you may need a hybrid TTS/accesskit-based approach to cover all bases.

Also, not *everything* needs an audio cue. Some events might work just as well with a speech alert (E.g. gaining/losing a new contact.)

The layers approach sounds good. You could also do something where all objects within a certain range have cues regardless of layer, and more distant targets need to be specifically selected/navigated toward, either automatically or with some sort of sonar ping on their location heard regardless of range.

Looks like a neat project. A space-based audio game is something I've wanted to do for years. I'll definitely keep tabs on this.

**Comment by bobbens** [comment]
Rust has been merged which means the first step towards resolving this has been taken. When the toolkit is rewritten in Rust, it should have accessibility in mind.

**Comment by zersiax** [comment]
Curious to see where this will go in the future, feel free to run questions by me (screen reader user / developer / gamer)  if there's anything I can clarify/answer
